djconnectwise/models.py

```diff
@@ -894,6 +1013,292 @@ def close(self, *args, **kwargs):
+            start_of_day = calendar.get_day_hours(True, day_of_week)
+            start = start.replace(
+                hour=start_of_day.hour,
+                minute=start_of_day.minute
+            )
+            days = 0
+
+        start_date = datetime.timedelta(hours=start.hour,
+                                        minutes=start.minute)
+
+        # Add minutes waiting to account for non-escalate time
+        sla_minutes = (sla_hours * 60) + self.minutes_waiting
+
+        end_of_day = calendar.get_day_hours(False, day_of_week)
+        end_of_day = datetime.timedelta(hours=end_of_day.hour,
+                                        minutes=end_of_day.minute)
+        first_day_minutes = (end_of_day - start_date).total_seconds() / 60
+
+        # If created outside of work hours, take no minutes off and start
+        # taking time off next work day
+        minutes = first_day_minutes if first_day_minutes >= 0 else 0
+        sla_minutes = sla_minutes - minutes
+
+        # Advance day by day, reducing the sla_minutes by the time in
+        # each working day.
+        # When minutes goes below zero, add the amount of time left on it
+        # that day to the start time of that day, giving you the due date
+        while sla_minutes >= 0:
+            day_of_week = (day_of_week + 1) % 7
+            days += 1
+
+            start_of_day = calendar.get_day_hours(True, day_of_week)
+            if start_of_day:
+                start_of_day = datetime.timedelta(hours=start_of_day.hour,
+                                                  minutes=start_of_day.minute)
+                end_of_day = calendar.get_day_hours(False, day_of_week)
+                end_of_day = datetime.timedelta(hours=end_of_day.hour,
+                                                minutes=end_of_day.minute)
+            else:
+                # This is a day with no hours, continue to next day
+                continue
+
+            minutes = (end_of_day - start_of_day).total_seconds() / 60
+            sla_minutes = sla_minutes - minutes
+
+        # sla_minutes went below zero so we know that day is the expiry
+        # Add the minutes back to sla_minutes and add sla minutes to the
+        # start of that day
+        sla_minutes = sla_minutes + minutes
+        expiry_date = start + datetime.timedelta(days=days)
+
+        if days == 0:
+            sla_expire_time = start + datetime.timedelta(minutes=sla_minutes)
+            expiry_date = expiry_date.replace(
+                hour=sla_expire_time.hour, minute=sla_expire_time.minute)
+        else:
+            sla_expire_time = calendar.get_day_hours(True, day_of_week)
+            expiry_date = expiry_date.replace(
+                hour=sla_expire_time.hour, minute=sla_expire_time.minute) + \
+                datetime.timedelta(minutes=sla_minutes)
+
+        self.sla_expire_date = expiry_date.astimezone(tz=timezone.utc)
+
+    def _get_sla_time(self, start, end, calendar):
```

Comment:
This method doesn't refer to `self` anywhere. Given that `calendar` is a parameter, maybe this would be better on the Calendar model?